### ISSUE TITLE
explicit/FoxGlynn.run() does not halt for certain parameters

### ISSUE THREAD

**Issue by muhrik** [prismmodelchecker/prism#72]
Hi!
I am currently working on implementing arbitrary precision-version of FoxGlynn using BigDecimal.
It is going along fine. I have a testing main method to compare results and performance with the original double-precision FoxGlynn. For certain parameters, this main method runs indefinitely. 
I have found out the problem lies in the original double-precision FoxGlynn.

**_INPUT:_**
```
double q = 4, time = 3; // q*time = q_tmax (the time parameter)
double uf = 1.0e-300; // underflow
double of = 1.0e-300; // overflow
double acc = 1.0e-25; // accuracy (of the result)
new FoxGlynn(q*time, uf, of, acc);
```
**_PROBLEM DESCRIPTION:_**
FoxGlynn has two approaches to computing results.
1. "Naive" approach is chosen if the first parameter q_tmax < 400. No further parameter checking is done.
2. "Actual" FoxGlynn is chosen otherwise. This approach first checks whether accuracy is not more precise than it can handle.
```
if (accuracy < 1e-10) {
   throw new PrismException("Overflow: Accuracy is smaller than Fox Glynn can handle (must be at least 1e-10).");
}
```

Since my input q_tmax = 12 is handled by the "naive" approach, the accuracy is not checked.
Through debugging, I found that this loop runs infinitely.
```
k = 1;
do {
    lastval *= q_tmax / k; // invariant: lastval = q_tmax^k / k!
    accum += lastval;
    w.add(lastval * expcoef);
    k++;
} while (accum < desval);
```
Eventually, the computed lastval increments get so small that adding them to accum does nothing.
Here is a debugging screenshot.
![screenshot from 2018-03-31 12-00-55](https://user-images.githubusercontent.com/32157431/38162268-79529d58-34de-11e8-97ca-5e7c4b68a3bd.png)

**_WHY THIS MATTERS:_**
FoxGlynn is used for many explicit engine computations within CTMCModelChecker.
While some methods have hardcoded termination epsilon (which is passed as accuracy to FoxGlynn), some do not. For example, I believe CTMCModelChecker.computeInstantaneousRewards uses termination epsilon from current PRISM settings. 
Say that Bob wants to compute instantaneous rewards for some CTMC model for very high precision 1.0e-30 using the explicit engine. Bob knows that such high precision computation will take a lot longer to finish, so he is prepared to wait a while. Sadly, Bob would have to wait forever (or until the w vector eats up the entire available memory, throwing java.lang.OutOfMemoryError: Java heap space).

**_PROPOSED SOLUTION:_**
Have both approaches check input parameters for too small accuracy. Perhaps even back in the FoxGlynn constructor. The "too small accuracy" problem is mostly related to double precision rather than the used approach.

**Comment by muhrik** [comment]
**_USER-PERSPECTIVE EXAMPLE:_**
Consider the following model.
```
ctmc

rate r = 405;

rewards
	s=0 : 1;
	s=1 : 50;
endrewards

module m

	s : [0..1] init 0;

	[] s=0 -> r: (s'=1);
	[] s=1 -> r: (s'=0);

endmodule

```
We verify property `R=? [I=100]` using the `explicit` engine for termination epsilon `1.0E-15`.

1. For `r = 405` the computation throws the following exception, correctly recognizing the precision as too much to handle.
`PrismException("Overflow: Accuracy is smaller than Fox Glynn can handle (must be at least 1e-10).")`
2. For `r = 1` the computation pretends to be able to handle this, but it can't.
The computation runs in an infinite loop until it runs out of memory.

**Comment by davexparker** [comment]
Thanks for the very helpful error report @muhrik. We've been discussing this today. I think the proposed solution of catching the "too small accuracy" scenario more consistently is going to b
e the best way to go.

**Comment by davexparker** [comment]
Fixed by #76.
